internal/provider/data_source_packer_image.go

```diff
@@ -111,9 +121,49 @@ func dataSourcePackerImageRead(ctx context.Context, d *schema.ResourceData, meta
 
 	log.Printf("[INFO] Reading HCP Packer registry (%s) [project_id=%s, organization_id=%s, iteration_id=%s]", bucketName, loc.ProjectID, loc.OrganizationID, iterationID)
 
-	iteration, err := clients.GetIterationFromId(ctx, client, loc, bucketName, iterationID)
-	if err != nil {
-		return diag.FromErr(err)
+	var iteration *packermodels.HashicorpCloudPackerIteration
+	var err error
+
+	if iterID, ok := iterationID.(string); ok && iterID != "" {
+		iteration, err = clients.GetIterationFromId(
+			ctx,
+			client,
+			loc,
+			bucketName,
+			iterID)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	var channel *packermodels.HashicorpCloudPackerChannel
+
+	if chanSlug, ok := channelName.(string); ok && chanSlug != "" {
+		channel, err = clients.GetPackerChannelBySlug(
+			ctx,
+			client,
+			loc,
+			bucketName,
+			chanSlug)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	var diags diag.Diagnostics
+
+	if channel != nil && iteration != nil {
+		return diag.FromErr(fmt.Errorf(
+			"iteration mismatch: channel %s's iteration (%s) is different from the explicitely specified iteration: %s",
+			channel.Slug,
+			channel.Iteration.ID,
+			iteration.ID))
+	}
+
+	// Assuming we passed the above check, the rest of the channel is not
+	// used after that,
+	if channel != nil {
+		iteration = channel.Iteration
 	}
```

Comment:
By using the helper function you can get rid of this as we will already have the iteration information we are looking for.